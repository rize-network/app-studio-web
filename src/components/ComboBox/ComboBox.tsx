import React from 'react';
import { View } from 'src/components';
export const ComboBox = () => { return <View><View/></View>; };