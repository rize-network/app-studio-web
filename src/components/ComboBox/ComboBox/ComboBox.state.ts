import { useState } from 'react';

export const useComboBoxState = () => {
  const [isHovered, setIsHovered] = useState(false);
  return { isHovered, setIsHovered };
};
