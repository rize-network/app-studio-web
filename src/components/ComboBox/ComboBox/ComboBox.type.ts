export type Styles = {
  [key: string]: React.CSSProperties;
};

export type ComboBoxStyles = {
  container?: React.CSSProperties;
  label?: React.CSSProperties;
  dropdown?: React.CSSProperties;
  item?: React.CSSProperties;
};
