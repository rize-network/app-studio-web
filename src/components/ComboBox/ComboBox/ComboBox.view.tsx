import React from 'react';
import { View } from 'src/components';
export const ComboBoxView = () => { return <View></View>; };