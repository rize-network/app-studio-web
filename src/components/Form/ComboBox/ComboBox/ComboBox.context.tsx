import { createContext, useContext } from 'react';
interface ComboBoxContextType {
  openDropdowns: { [key: string]: boolean };
  toggleDropdown: (id: string) => void;
}
const ComboBoxContext = createContext<ComboBoxContextType>(null!);
export const useComboBoxContext = () => useContext(ComboBoxContext);
export default ComboBoxContext;
