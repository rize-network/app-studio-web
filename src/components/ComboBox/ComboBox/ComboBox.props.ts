import { ReactNode, Dispatch, SetStateAction } from 'react';
import { ComboBoxStyles } from './ComboBox.type';

export interface ComboBoxProps {
  items: ComboBoxItem[];
  onSelect?: (item: ComboBoxItem) => void;
  searchEnabled?: boolean;
  left?: ReactNode;
  right?: ReactNode;
  label?: string;
  showTick?: boolean;
  placeholder?: string;
  styles?: ComboBoxStyles;
}

export interface ComboBoxItem {
  label: string;
  value: string;
  icon?: ReactNode;
}

export interface ComboBoxStateActions {
  filteredItems: ComboBoxItem[];
  setFilteredItems: Dispatch<SetStateAction<ComboBoxItem[]>>;
  selectedItem: ComboBoxItem;
  setSelectedItem: Dispatch<SetStateAction<ComboBoxItem>>;
  highlightedIndex: number;
  setHighlightedIndex: Dispatch<SetStateAction<number>>;
  searchQuery: string;
  setSearchQuery: Dispatch<SetStateAction<string>>;
}

export interface ComboBoxViewProps extends ComboBoxProps, ComboBoxStateActions {
  // You can add additional properties here if needed
}
