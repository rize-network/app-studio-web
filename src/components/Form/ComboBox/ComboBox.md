# ComboBox

A customizable dropdown list with optional search functionality.

### **Import**

```tsx static
import { ComboBox } from '@app-studio/web';
```

### **Default**

```tsx
import React from 'react';
import { ComboBox } from '../ComboBox';

export const DefaultDemo = () => {
  const items = [
    {
      value: 'next.js',
      label: 'Next.js',
    },
    {
      value: 'sveltekit',
      label: 'SvelteKit',
    },
    {
      value: 'nuxt.js',
      label: 'Nuxt.js',
    },
  ];
  return <ComboBox id="default" name="default" items={items} />;
};
```

### **onSelect**

"Optional callback function triggered when an item is selected."

```tsx
import React from 'react';
import { ComboBox } from '../ComboBox';
import { Vertical } from '../../../Layout';
import { MessageLayout, showMessage } from '../../..';

export const OnSelectDemo = () => {
  const items = [
    {
      value: 'next.js',
      label: 'Next.js',
    },
    {
      value: 'sveltekit',
      label: 'SvelteKit',
    },
    {
      value: 'nuxt.js',
      label: 'Nuxt.js',
    },
    {
      value: 'remix',
      label: 'Remix',
    },
    {
      value: 'astro',
      label: 'Astro',
    },
  ];
  return (
    <Vertical gap={15}>
      <ComboBox
        id="onSelect"
        items={items}
        onSelect={(item) => {
          showMessage('success', `Item selected: ${item.label}`);
        }}
      />
      <MessageLayout position="bottomRight" />
    </Vertical>
  );
};
```

### **searchEnabled**

"Flag to enable search functionality within the ComboBox."

```tsx
import React from 'react';
import { ComboBox } from '../ComboBox';

export const SearchEnabledDemo = () => {
  const items = [
    {
      value: 'next.js',
      label: 'Next.js',
    },
    {
      value: 'sveltekit',
      label: 'SvelteKit',
    },
    {
      value: 'nuxt.js',
      label: 'Nuxt.js',
    },
    {
      value: 'remix',
      label: 'Remix',
    },
    {
      value: 'astro',
      label: 'Astro',
    },
  ];
  return <ComboBox id="searchEnabled" items={items} searchEnabled={false} />;
};
```

### **left**

"Optional element to display on the left side of the ComboBox."

```tsx
import React from 'react';
import { Badge } from '../../..';
import { ComboBox } from '../ComboBox';

export const LeftDemo = () => {
  const items = [
    {
      value: 'enhancement',
      label: 'Enhancement',
    },
    {
      value: 'bug',
      label: 'Bugs',
    },
  ];

  return (
    <ComboBox
      id="left"
      name="left"
      items={items}
      left={<Badge content="Status" />}
    />
  );
};
```

### **right**

"Optional element to display on the right side of the ComboBox."

```tsx
import React from 'react';
import { Badge } from '../../..';
import { ComboBox } from '../ComboBox';

export const RightDemo = () => {
  const items = [
    {
      value: 'enhancement',
      label: 'Enhancement',
    },
    {
      value: 'bug',
      label: 'Bugs',
    },
  ];

  return (
    <ComboBox id="right" items={items} right={<Badge content="Status" />} />
  );
};
```

### **label**

"Optional label to describe the purpose of the ComboBox."

```tsx
import React from 'react';
import PlusSvg from 'src/components/Svg/Plus';
import { ComboBox } from '../ComboBox';

export const LabelDemo = () => {
  const items = [
    {
      value: 'todo',
      label: 'To Do',
    },
    {
      value: 'inprogress',
      label: 'In Progress',
    },
  ];
  return (
    <ComboBox
      id="label"
      name="label"
      label="Status"
      items={items}
      left={<PlusSvg size={12} />}
    />
  );
};
```

### **showTick**

"Flag to show a tick mark next to selected items."

```tsx
import React from 'react';
import { ComboBox } from '../ComboBox';

export const ShowTickDemo = () => {
  const statuses = [
    {
      value: 'backlog',
      label: 'Backlog',
    },
    {
      value: 'todo',
      label: 'Todo',
    },
    {
      value: 'in progress',
      label: 'In Progress',
    },
    {
      value: 'done',
      label: 'Done',
    },
    {
      value: 'canceled',
      label: 'Canceled',
    },
  ];
  return <ComboBox id="showTick" items={statuses} showTick={false} />;
};
```

### **placeholder**

"Text to display when nothing has been selected in the ComboBox."

```tsx
import React from 'react';
import { ComboBox } from '../ComboBox';

export const PlaceholderDemo = () => {
  const items = [
    {
      value: 'todo',
      label: 'To Do',
    },
    {
      value: 'inprogress',
      label: 'In Progress',
    },
  ];
  return (
    <ComboBox
      id="placeholder"
      name="placeholder"
      placeholder="Select..."
      searchEnabled={false}
      items={items}
    />
  );
};
```

### **styles**

"Optional custom styles to apply to the ComboBox."

```tsx
import React from 'react';
import { ComboBox } from '../ComboBox';

export const StylesDemo = () => {
  const items = [
    {
      value: 'todo',
      label: 'To Do',
    },
    {
      value: 'inprogress',
      label: 'In Progress',
    },
  ];
  return (
    <ComboBox
      id="styles"
      items={items}
      searchEnabled={false}
      styles={{
        container: {
          boxShadow:
            'rgba(50, 50, 93, 0.25) 0px 50px 100px -20px, rgba(0, 0, 0, 0.3) 0px 30px 60px -30px, rgba(10, 37, 64, 0.35) 0px -2px 6px 0px inset',
        },
        item: {
          margin: '10px',
        },
        dropdown: {
          boxShadow:
            'rgba(50, 50, 93, 0.25) 0px 50px 100px -20px, rgba(0, 0, 0, 0.3) 0px 30px 60px -30px, rgba(10, 37, 64, 0.35) 0px -2px 6px 0px inset',
        },
        text: {
          borderBottom: '1px solid transparent',
        },
      }}
    />
  );
};
```

### **searchPlaceholder**

"Placeholder text for the search input when search is enabled."

```tsx
import React from 'react';
import { ComboBox } from '../ComboBox';

export const SearchPlaceholderDemo = () => {
  const items = [
    {
      value: 'todo',
      label: 'To Do',
    },
    {
      value: 'inprogress',
      label: 'In Progress',
    },
  ];
  return (
    <ComboBox
      id="searchPlaceholder"
      placeholder="Select..."
      searchPlaceholder="FilterStatus.."
      items={items}
    />
  );
};
```
