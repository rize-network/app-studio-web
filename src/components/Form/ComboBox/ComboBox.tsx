import React from 'react';
import { ComboBoxProps } from './ComboBox/ComboBox.props';
import { useComboBoxState } from './ComboBox/ComboBox.state';
import ComboBoxView from './ComboBox/ComboBox.view';

const ComboBoxComponent: React.FC<ComboBoxProps> = ({
  items,
  placeholder,
  searchPlaceholder,
  ...props
}) => {
  const state = useComboBoxState(items, placeholder, searchPlaceholder);

  return <ComboBoxView items={items} {...state} {...props} />;
};

export const ComboBox = ComboBoxComponent;
