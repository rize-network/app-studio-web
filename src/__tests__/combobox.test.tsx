import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { ComboBox } from '../components/Form/ComboBox/ComboBox';
import renderer from 'react-test-renderer';

const mockItems = [
  { label: 'Option 1', value: '1' },
  { label: 'Option 2', value: '2' },
  { label: 'Option 3', value: '3' },
];

describe('ComboBox', () => {
  test('renders Checkbox component', () => {
    render(<ComboBox id="combo" items={mockItems} />);
    const CheckboxElement = screen.getByRole('combobox');
    expect(CheckboxElement).toBeInTheDocument();
  });

  test('renders with placeholder text', () => {
    render(
      <ComboBox id="combo" placeholder="Select an option" items={mockItems} />
    );
    expect(screen.getByRole('combobox')).toHaveTextContent('Select an option');
  });

  test('does not show tick next to selected item if showTick is false', () => {
    render(<ComboBox id="combo" items={mockItems} showTick={false} />);
    fireEvent.click(screen.getByRole('combobox'));
    fireEvent.click(screen.getByText('Option 1'));
    expect(screen.queryByRole('img')).not.toBeInTheDocument();
  });
});

test('Checkbox to match snapshot', () => {
  const tree = renderer
    .create(<ComboBox id="combo" items={mockItems} showTick={false} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
