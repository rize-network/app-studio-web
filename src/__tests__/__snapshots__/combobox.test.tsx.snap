// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox to match snapshot 1`] = `
<div
  className="sc-bdfCDU foBlXZ"
  id="combo"
  role="combobox"
>
  <div
    className="sc-bdfCDU dRoCvi"
  >
    <div
      className="sc-bdfCDU eTWVak"
    >
      <div
        className="sc-bdfCDU jwEnfw"
        onClick={[Function]}
      >
        <div
          className="sc-bdfCDU gVoSlS"
          role="text"
        >
          Option 1
        </div>
      </div>
    </div>
  </div>
</div>
`;
